drivers/sensor/bmp581/bmp581.c

```diff
@@ -0,0 +1,565 @@
+			  struct bmp581_data *drv)
+{
+	int ret = 0;
+
+	if (osr != NULL) {
+		uint8_t oversampling = osr->val1;
+		uint8_t press_en = osr->val2 != 0; /* if it is not 0 then pressure is enabled */
+		uint8_t osr_val = 0;
+
+		ret = reg_read(BMP5_REG_OSR_CONFIG, &osr_val, 1, drv);
+		if (ret == BMP5_OK) {
+			switch (chan) {
+			case SENSOR_CHAN_ALL:
+				osr_val = BMP5_SET_BITS_POS_0(osr_val, BMP5_TEMP_OS, oversampling);
+				osr_val = BMP5_SET_BITSLICE(osr_val, BMP5_PRESS_OS, oversampling);
+				osr_val = BMP5_SET_BITSLICE(osr_val, BMP5_PRESS_EN, press_en);
+				break;
+			case SENSOR_CHAN_PRESS:
+				osr_val = BMP5_SET_BITSLICE(osr_val, BMP5_PRESS_OS, oversampling);
+				osr_val = BMP5_SET_BITSLICE(osr_val, BMP5_PRESS_EN, press_en);
+				break;
+			case SENSOR_CHAN_AMBIENT_TEMP:
+				osr_val = BMP5_SET_BITS_POS_0(osr_val, BMP5_TEMP_OS, oversampling);
+				break;
+			default:
+				ret = -ENOTSUP;
+				break;
+			}
+
+			if (ret == BMP5_OK) {
+				ret = reg_write(BMP5_REG_OSR_CONFIG, &osr_val, 1, drv);
+				get_osr_odr_press_config(&drv->osr_odr_press_config, drv);
+			}
+		}
+	} else {
+		ret = BMP5_E_NULL_PTR;
+	}
+
+	return ret;
+}
+
+static int set_odr_config(const struct sensor_value *odr, struct bmp581_data *drv)
+{
+	int ret = 0;
+
+	if (odr != NULL) {
+		uint8_t odr_val = 0;
+
+		ret = reg_read(BMP5_REG_ODR_CONFIG, &odr_val, 1, drv);
+		if (ret != BMP5_OK) {
+			return ret;
+		}
+		odr_val = BMP5_SET_BITSLICE(odr_val, BMP5_ODR, odr->val1);
+		ret = reg_write(BMP5_REG_ODR_CONFIG, (const uint8_t *)&odr_val, 1, drv);
+		get_osr_odr_press_config(&drv->osr_odr_press_config, drv);
+	} else {
+		ret = BMP5_E_NULL_PTR;
+	}
+
+	return ret;
+}
+
+static int soft_reset(struct bmp581_data *drv)
+{
+	int ret = 0;
+	const uint8_t reset_cmd = BMP5_SOFT_RESET_CMD;
+	uint8_t int_status = 0;
+
+	ret = reg_write(BMP5_REG_CMD, &reset_cmd, 1, drv);
+
+	if (ret == BMP5_OK) {
+		k_usleep(BMP5_DELAY_US_SOFT_RESET);
+		if (ret == BMP5_OK) {
+			ret = reg_read(BMP5_REG_INT_STATUS, &int_status, 1, drv);
+			if (ret == BMP5_OK) {
+				if (int_status & BMP5_INT_ASSERTED_POR_SOFTRESET_COMPLETE) {
+					ret = BMP5_OK;
+				} else {
+					ret = BMP5_E_POR_SOFTRESET;
+				}
+			} else {
+				ret = BMP5_E_NULL_PTR;
+			}
+		}
+	} else {
+		LOG_DBG("Failed perform soft-reset.");
+	}
+
+	return ret;
+}
+
+static int bmp581_sample_fetch(const struct device *dev, enum sensor_channel chan)
+{
+	struct bmp581_data *drv = (struct bmp581_data *)dev->data;
+	uint8_t data[6];
+	int ret = 0;
+
+	if (chan != SENSOR_CHAN_ALL) {
+		return -ENOTSUP;
+	}
+
+	ret = reg_read(BMP5_REG_TEMP_DATA_XLSB, data, 6, drv);
+	if (ret == BMP5_OK) {
+		int32_t raw_temperature =
+			(int32_t)((uint32_t)(data[2] << 16) | (uint16_t)(data[1] << 8) | data[0]);
+		/*
+		 * Division by 2^16(whose equivalent value is 65536) is performed to get temperature
+		 * data in deg C
+		 */
+
+		sensor_value_from_double(&drv->last_sample.temperature,
+								(raw_temperature / 65536.0));
```

Comment:
Don't need to use floating point to divide by a power of 2

Reply:
You lose the decimal places when you divide an int by an int.

Reply:
Some platforms in Zephyr don't have a hardware FPU so we try to avoid floating point operations if possible. I think you can do something like this:
```c
/* use the integer part directly (may need to sign extend) */
drv->last_sample.temperature.val1 = data[2];

/* convert the fractional part to decimal */
drv->last_sample.temperature.val2 = 1000000 * ([data[1] << 8 | data[0]);
```